### PR TITLE
Add support for mixed meshes using libCEED.

### PR DESCRIPTION
**Pull request by YohannDudouit**
This PR adds PA and MF support on GPU for mixed finite element spaces using libCEED.

~~I would recommend reviewing this PR after https://github.com/mfem/mfem/pull/2912 and https://github.com/mfem/mfem/pull/2952.~~

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2953](https://github.com/mfem/mfem/pull/2953) | @YohannDudouit | @tzanio | @acfisher + @v-dobrev + @artv3 | 6/4/22 | 7/12/22 | 7/28/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by YohannDudouit** [comment]
I would recommend reviewing this PR after #2912 and #2952.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by YohannDudouit** [comment]
We did a test with @psocratis , this PR also adds support for p-adaptivity.

File: CHANGELOG [hunk 1]
```diff
@@ -44,6 +44,13 @@ Version 4.4.1 (development)
 
 - Added support for mixed meshes and pyramids in GSLIB-FindPoints.
 
+- Added support for partial assembly and fully matrix-free operators on
+  mixed meshes (different element topologies and p-adaptivity) through
+  libCEED. Include device support, e.g. NVIDIA and AMD GPUs.
+  The p-adaptivity is currently limited by MFEM capabilities, i.e. 2D serial
+  meshes. All mixed element topologies are supported in serial and parallel:
+  segment, triangle, square, tetrahedron, cube, prism, and pyramid.
+
```
Comment (artv3) [inline CHANGELOG:53]: Which version of libceed do need for this support?

File: fem/ceed/interface/coefficient.hpp [hunk 2]
```diff
@@ -22,6 +21,8 @@
 #include "util.hpp"
 #include "ceed.hpp"
 
+#ifdef MFEM_USE_CEED
```
Comment (acfisher) [inline fem/ceed/interface/coefficient.hpp:24]: Is there a reason this isn't above the include statements?

File: fem/ceed/interface/restriction.cpp [hunk 3]
```diff
@@ -31,77 +31,189 @@ static void InitNonTensorRestriction(const mfem::FiniteElementSpace &fes,
...
+static void InitRestrictionWithIndicesImpl(
+   const mfem::FiniteElementSpace &fes,
+   int nelem,
+   const int* indices,
+   Ceed ceed, CeedElemRestriction *restr)
+{
+   const mfem::FiniteElement *fe = fes.GetFE(indices[0]);
+   const mfem::TensorBasisElement * tfe =
+      dynamic_cast<const mfem::TensorBasisElement *>(fe);
+   if (tfe)
```
Comment (acfisher) [inline fem/ceed/interface/restriction.cpp:171]: Similar style comment about the if logic as above.

File: tests/unit/ceed/test_ceed.cpp [hunk 4]
```diff
@@ -427,14 +514,35 @@ TEST_CASE("CEED mass & diffusion", "[CEED]")
    auto pb = GENERATE(Problem::Mass,Problem::Diffusion,Problem::MassDiffusion,
                       Problem::VectorMass,Problem::VectorDiffusion);
    auto order = GENERATE(1);
-   auto mesh = GENERATE("../../data/inline-quad.mesh","../../data/inline-hex.mesh",
+   auto mesh = GENERATE("../../data/inline-quad.mesh",
```
Comment (acfisher) [inline tests/unit/ceed/test_ceed.cpp:517]: I think this is the first time I have seen the Catch generators in action. I like it quite a bit.

**Review by acfisher — APPROVED** [review]

File: fem/ceed/interface/coefficient.hpp (hunk 2 above)
Comment (YohannDudouit) [reply fem/ceed/interface/coefficient.hpp]: Good catch, done in [c11576a](https://github.com/mfem/mfem/pull/2953/commits/c11576a686e107a11e97d5430ac22aa817c3b63b).

File: fem/ceed/interface/restriction.cpp [hunk 5]
```diff
@@ -31,77 +31,189 @@ static void InitNonTensorRestriction(const mfem::FiniteElementSpace &fes,
...
-                           Ceed ceed, CeedElemRestriction *restr)
+static void InitRestrictionImpl(const mfem::FiniteElementSpace &fes,
+                                Ceed ceed, CeedElemRestriction *restr)
 {
    const mfem::FiniteElement *fe = fes.GetFE(0);
    const mfem::TensorBasisElement * tfe =
       dynamic_cast<const mfem::TensorBasisElement *>(fe);
-   MFEM_VERIFY(tfe, "invalid FE");
-   const mfem::Array<int>& dof_map = tfe->GetDofMap();
+   if (tfe)
```
[the remark this answers is not in the document]
Comment (YohannDudouit) [reply fem/ceed/interface/restriction.cpp]: I agree it's better, done in [c11576a](https://github.com/mfem/mfem/pull/2953/commits/c11576a686e107a11e97d5430ac22aa817c3b63b).

**Review by tzanio — APPROVED** [review]
Thanks for adding this @YohannDudouit !

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by YohannDudouit** [comment]
@tzanio The bug has been found by @mlstowell , we're good to remerge in next once CI passes.

**Comment by tzanio** [comment]
Excellent, thanks guys!

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**